cibuildwheel/platforms/android.py

```diff
@@ -0,0 +1,429 @@
+import importlib.util
+import os
+import platform
+import re
+import shlex
+import shutil
+import subprocess
+from collections.abc import Generator
+from contextlib import contextmanager
+from dataclasses import dataclass
+from pathlib import Path
+from pprint import pprint
+from typing import Any
+
+from build import ProjectBuilder
+from filelock import FileLock
+
+from .. import errors
+from ..architecture import Architecture, arch_synonym, native_platform
+from ..frontend import BuildFrontendConfig, get_build_frontend_extra_flags, parse_config_settings
+from ..logger import log
+from ..options import BuildOptions, Options
+from ..selector import BuildSelector
+from ..typing import PathOrStr
+from ..util import resources
+from ..util.cmd import call, shell
+from ..util.file import CIBW_CACHE_PATH, copy_test_sources, download, move_file
+from ..util.helpers import prepare_command
+from ..util.packaging import find_compatible_wheel
+from ..venv import constraint_flags, virtualenv
+
+
+def android_triplet(identifier: str) -> str:
+    return {
+        "arm64_v8a": "aarch64-linux-android",
+        "x86_64": "x86_64-linux-android",
+    }[parse_identifier(identifier)[1]]
+
+
+def parse_identifier(identifier: str) -> tuple[str, str]:
+    match = re.fullmatch(r"cp(\d)(\d+)-android_(.+)", identifier)
+    if not match:
+        msg = f"invalid Android identifier: '{identifier}'"
+        raise ValueError(msg)
+    major, minor, arch = match.groups()
+    return (f"{major}.{minor}", arch)
+
+
+@dataclass(frozen=True)
+class PythonConfiguration:
+    version: str
+    identifier: str
+    url: str
+
+    @property
+    def arch(self) -> str:
+        return parse_identifier(self.identifier)[1]
+
+
+def all_python_configurations() -> list[PythonConfiguration]:
+    return [PythonConfiguration(**item) for item in resources.read_python_configs("android")]
+
+
+def get_python_configurations(
+    build_selector: BuildSelector, architectures: set[Architecture]
+) -> list[PythonConfiguration]:
+    return [
+        c
+        for c in all_python_configurations()
+        if c.arch in architectures and build_selector(c.identifier)
+    ]
+
+
+def shell_prepared(command: str, build_options: BuildOptions, env: dict[str, str]) -> None:
+    shell(
+        prepare_command(command, project=".", package=build_options.package_dir),
+        env=env,
+    )
+
+
+def before_all(options: Options, python_configurations: list[PythonConfiguration]) -> None:
+    before_all_options = options.build_options(python_configurations[0].identifier)
+    if before_all_options.before_all:
+        log.step("Running before_all...")
+        shell_prepared(
+            before_all_options.before_all,
+            before_all_options,
+            before_all_options.environment.as_dictionary(os.environ),
+        )
+
+
+@dataclass
+class Builder:
```

Comment:
I'm not entirely sure I see the benefit of introducing a class here. It's inconsistent with the implementation approach of other backends; the module is already providing the "unit of code" architecture; and it's not being passed outside this module as a convenience. The one possible reason I can see is that it avoids passing around some variables - but that's clearly causing the linter some issues.

Reply:
OK, I'll remove the class. Out of interest, what linter issues did you mean?

Reply:
Probably PyLint issues, which only show up as annotations, and don't fail the build. `nox -s pylint` runs them locally.